BUG: `line terminator` and '\n to \r\n' problem in Windows(Issue #20353)

deflatSOCO opened the pull request:
- [ ] closes #20353
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

gfyoung commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -147,7 +147,7 @@ MultiIndex
 ^^^
 
--
+- (Only for Windows) Bug in :meth:`DataFrame.to_csv`, in which all `\n`s are converted to `\r\n` (:issue:`20353`)
```
I would move the Windows portion as follows:

`Bug in :meth:`DataFrame.to_csv`, in which all `\n`s are converted to `\r\n` on Windows (:issue:`20353`)`

gfyoung commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -285,3 +285,62 @@ def test_to_csv_string_array_utf8(self):
+                b'int,str_crlf\n' \
+                b'1,abc\n' \
+                b'2,"d\r\nef"\n' \
+                b'3,"g\r\nh\r\n\r\ni"\n'
```
Use parentheses instead of slashes for multi-line strings.

gfyoung commented on pandas/io/common.py:
```diff
@@ -394,13 +395,13 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
         if compat.PY2:
             # Python 2
-            f = open(path_or_buf, mode)
+            f = io.open(path_or_buf, mode, newline='\n')
```
Why do we need to use `io.open` for this one?

deflatSOCO replied:
I referenced [this stackoverflow article](https://stackoverflow.com/questions/34148441/python-2-7-newline-in-function), but it was not appropreate as all CI tests failed in py2 environment.
I will fix it.

pep8speaks commented:
Hello @deflatSOCO! Thanks for updating the PR.





 - In the file [`pandas/tests/io/formats/test_to_csv.py`](https://github.com/pandas-dev/pandas/blob/67f010a0ec53195b9c77ebf840f5a85caf174995/pandas/tests/io/formats/test_to_csv.py), following are the PEP8 issues :

> [Line 310:27](https://github.com/pandas-dev/pandas/blob/67f010a0ec53195b9c77ebf840f5a85caf174995/pandas/tests/io/formats/test_to_csv.py#L310): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator

bibenb1 commented on pandas/io/common.py:
```diff
@@ -394,13 +394,15 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
         elif encoding:
             # Python 3 and encoding
-            f = open(path_or_buf, mode, encoding=encoding)
+            f = open(path_or_buf, mode, encoding=encoding, newline='\n')
```
@deflatSOCO Is there a specific reason for not making newline a function parameter? I would suggest that you make _newline_ a `_get_handle` parameter and default it to *None* then let `open` handle it from there.
Then update `formats.csvs.CSVFormatter.save` to pass _self.line_terminator_ to the function

chris-b1 commented:
Willing to be convinced otherwise but I don't think the default should change to `\n` on Windows.  It's an annoying convention, but convention all the same.

gfyoung commented:
@chris-b1 : The issue more relates to an inconsistency, as <a href="https://github.com/pandas-dev/pandas/issues/20353#issuecomment-392301505">here</a>.

chris-b1 commented:
Right, and I'm good with fixing the original bug (`\r\n` terminator when `\n` explicitly passed), but this PR would, if I'm understanding correctly, also change the default to `\n` when no line terminator is passed.

gfyoung commented:
> but this PR would, if I'm understanding correctly, also change the default to \n when no line terminator is passed.

@chris-b1 : Ah, I see what you mean.  That is indeed correct based on the tests.

@deflatSOCO : Could you change it so that the default remains `\r\n` unless otherwise specified?

deflatSOCO commented:
@chris-b1 @gfyoung  Please let me convince.
[The current document](http://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.to_csv.html) says the default value of line_terminator in `to_csv` is **'\n'** no matter what OS is, so I just followed that in this PR.
I suppose that you mean that should be changed to '\r\n' only in Windows.

If so, the expected output should be like this. Is that right?
* code
```
data = pd.DataFrame({
    "integer":[1,2,3],
    "string_lf":["abc","d\nef","gh\r\ni"]
})
data.to_csv('test.txt', index=False)
print(pd.read_csv('test.csv'))
data.to_csv('test.csv', index=False)
print(pd.read_csv('test.csv'))
with open("test.csv", mode='rb') as f:
    print(f.read())
```
* current output(v0.23)
```
   integer  string_lf
0        1        abc
1        2    d\r\nef
2        3  gh\r\r\ni
b'integer,string_lf\r\n1,abc\r\n2,"d\r\nef"\r\n3,"gh\r\r\ni"\r\n'
```
* current output(this PR)
```
   integer  string_lf
0        1        abc
1        2      d\nef
2        3    gh\r\ni
b'integer,string_lf\n1,abc\n2,"d\nef"\n3,"gh\ni"\n'
```
* expected output
```
   integer  string_lf
0        1        abc
1        2      d\nef
2        3    gh\r\ni
b'integer,string_lf\r\n1,abc\r\n2,"d\nef"\r\n3,"gh\ni"\r\n'
```

gfyoung commented:
> @chris-b1 @gfyoung Please let me convince.
The current document says the default value of line_terminator in to_csv is '\n' no matter what OS is, so I just followed that in this PR.

@chris-b1 : Thoughts?  I think @deflatSOCO has a point here.

gfyoung commented:
@jreback @chris-b1 : Could you take a look at this?

chris-b1 commented:
I know the current docs say contrary, but my vote is that we continue defaulting to the platform default line terminator (`\r\n` on Windows) and updating the docs as such.

It preserves backwards compat - so if someone was relying on the `\r` being there, their code won't break, and is consistent with other programs on Windows that save csvs, including Excel, and python's csv writer.

```python
In [1]: import csv

In [3]: with open('tmp.csv', 'wt', newline='') as f:
   ...:     writer = csv.writer(f)
   ...:     writer.writerow(['a', 'b'])
   ...:     writer.writerow([1, 2])
   ...:     writer.writerow([3, 4])

In [4]: open('tmp.csv', 'rb').read()
Out[4]: b'a,b\r\n1,2\r\n3,4\r\n'
```

(again fully in support of merging this with bugfix for case with explicit ending)

jreback commented:
can u rebase

gfyoung commented on pandas/tests/frame/test_to_csv.py:
```diff
@@ -841,10 +842,10 @@ def test_to_csv_unicodewriter_quoting(self):
+        expected = ('"A","B"' + os.linesep +
+                    '1,"foo"' + os.linesep +
+                    '2,"bar"' + os.linesep +
+                    '3,"baz"' + os.linesep)
```
I wonder if we can make a testing utility for this.

jreback commented on pandas/tests/frame/test_to_csv.py:
```diff
@@ -865,31 +869,38 @@ def test_to_csv_index_no_leading_comma(self):
+            with open(path, mode='rb') as f:
+                assert f.read() == expected
+
+            df.to_csv(path, line_terminator='\n')
```
can you do separate ensure_cleans() for each case & put a comment for each case

jreback commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -286,21 +342,114 @@ def test_to_csv_string_array_utf8(self):
+                b'2,"d\nef"' + os_linesep +
+                b'3,"g\nh\n\ni"' + os_linesep
+            )
+
```
can you make separate ensure_cleans here

jreback commented on pandas/tests/io/formats/test_to_csv.py:
```diff
@@ -286,21 +342,114 @@ def test_to_csv_string_array_utf8(self):
         assert not sys.stdout.closed
 
+    @pytest.mark.xfail(
+        os.name == 'nt',
```
use ``compat.is_windows_platform()``

jreback commented on pandas/util/testing.py:
```diff
@@ -2795,3 +2795,10 @@ def skipna_wrapper(x):
+
+
+def convert_rows_list_to_csv_str(rows_list):
+    csv_str = ''
```
can you add a mini-doc-string

jreback commented on pandas/util/testing.py:
```diff
@@ -2795,3 +2795,10 @@ def skipna_wrapper(x):
+    csv_str = ''
+    for row in rows_list:
+        csv_str += row + os.linesep
+    return csv_str
```
can you use 

``return os.linesep.join(rows_list)`` ?

gfyoung commented on pandas/util/testing.py:
```diff
@@ -2801,3 +2801,24 @@ def skipna_wrapper(x):
     return skipna_wrapper
+
+
+def convert_rows_list_to_csv_str(rows_list):
```
Feedback was addressed in a715d9c.

gfyoung commented:
@deflatSOCO : Could you address the merge conflicts here?

@TomAugspurger @jreback : Can you have a look again at this?

deflatSOCO commented:
@TomAugspurger @jreback : Can you have a look again at this?

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -189,6 +189,7 @@ Other Enhancements
 
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+- :func:`DataFrame.to_csv` now uses :func:`os.linesep` rather than ``'\n'`` for the default line terminator.(:issue:`20353`)
```
can you elabortate on what if anything a user needs to do (e.g. when should they care about this)

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -731,6 +732,7 @@ MultiIndex
 I/O
 ^^^
 
+- Bug in :meth:`DataFrame.to_csv`, in which all `\n`s are converted to `\r\n` on Windows (:issue:`20353`)
```
is this a repeat?

jreback reviewed (changes requested):
looks ok to me. i would just request an exapanded whatsnew note. maybe a small sub-section on when a user should care and what they need to do about it. e.g. i write csv's on windows, do I need to do anything?what if I explicity pass ``line_terminator``

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -196,6 +196,158 @@ Other Enhancements
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:func:`DataFrame.to_csv` now uses :func:`os.linesep` rather than ``'\n'``
+ for the default line terminator(:issue:`20353`).
```
space after terminator

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -196,6 +196,158 @@ Other Enhancements
+:func:`DataFrame.to_csv` now uses :func:`os.linesep` rather than ``'\n'``
+ for the default line terminator(:issue:`20353`).
+- This change only affects when running on Windows, where ``'\r\n'`` was used for line terminator
+even when ``'\n'`` was passed in ``line_terminator``.
```
no need to use the leading '-', just write more in a paragraph style. First sentence is great, not sure I understand the 2nd and 3rd ones.

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -196,6 +196,158 @@ Other Enhancements
+
+.. code-block:: ipython
+
+  In [1]: import pandas as pd
```
you can skip the pd import

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -196,6 +196,158 @@ Other Enhancements
+
+  In [5]: with open("test.csv", mode='rb') as f:
+     ...:     print(f.read())
+  b'string_with_lf,string_with_crlf\r\nabc,abc\r\n"d\r\nef","d\r\r\nef"\r\n"g\r\nh
```
can make this simpler, just show [3] , [4]

jreback replied:
yes, the point is we want a small example that shows the most common case where a user would want to know what is changing. Too long and you lose the reader, too short and you don't know if the change applies to youl.

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -196,6 +196,158 @@ Other Enhancements
+
+  In [1]: import pandas as pd
+
+  In [2]: data = pd.DataFrame({
```
same as above

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -196,6 +196,158 @@ Other Enhancements
+     ...:     data.to_csv(f,index=False,line_terminator='\n')
+
+  In [7]: pd.read_csv("test2.csv")
+  Out[7]:
```
thanks for writing all this, but just show a single, most important change

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -196,6 +196,94 @@ Other Enhancements
+   ...: "string_with_lf":["a\nbc"],
+   ...: "string_with_crlf":["a\r\nbc"]
+   ...: })
+
```
can you use pep in these examples, e.g. put spaces after the commas

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -196,6 +196,94 @@ Other Enhancements
+
+In [3]: with open("test.csv", mode='rb') as f:
+   ...:     print(f.read())
+b'string_with_lf,string_with_crlf\r\n"a\r\nbc","a\r\r\nbc"\r\n'
```
I still not clear what these cases are , can you add a coment for example what 2/3 differe from 4/5

deflatSOCO commented:
@jreback I applied in commit e4badc4, and rebased. Please check them.

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -201,6 +201,96 @@ Backwards incompatible API changes
+b'string_with_lf,string_with_crlf\n"a\nbc","a\r\nbc"\n'
+
+
+New Behavior on Windows:
```
@gfyoung can you have a read over this. This should just point out the most common case / change that matters.

gfyoung replied:
Overall, this looks pretty good!  Left a few comments below.

gfyoung commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -201,6 +201,96 @@ Backwards incompatible API changes
+b'string_with_lf,string_with_crlf\n"a\nbc","a\r\nbc"\n'
+
+
+- On windows, the value of ``os.linesep`` is ``'\r\n'``,
```
windows --> Windows

gfyoung commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -201,6 +201,96 @@ Backwards incompatible API changes
+
+- By passing ``line_terminator`` explicitly, line terminator is set to that character.
+- The value of ``line_terminator`` only affects the line terminator of CSV,
+  so it does not change the value inside the data.
```
"it does not change the value inside the data" - what do you mean by that exactly?

deflatSOCO replied:
Suppose that `data` is like this (which is used in following codes)

||string_with_lf|string_with_crlf|
|:-|:-|:-|
|1|a\nbc|a\r\nbc|

As described in [2]&[3] of "Previous Behavior", if there are some `\n`s INSIDE the data, `to_csv()` changed the value and outputted CSV like this.

||string_with_lf|string_with_crlf|
|:-|:-|:-|
|1| a **\r** \nbc | a **\r** \r\nbc |

By this PR, it is fixed to output the following CSV.

||string_with_lf|string_with_crlf|
|:-|:-|:-|
|1| a\nbc | a\r\nbc |

gfyoung replied:
Gotcha.  Could you mention that in [2] and [3]?  What you're describing isn't explicitly mentioned.

gfyoung commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -201,6 +201,96 @@ Backwards incompatible API changes
+
+
+- As default value of ``line_terminator`` changes, just passing file object with ``newline='\n'`` does not set ``'\n'`` to line terminator.
+  Pass ``line_terminator='\n'`` explicitly.
```
I think we can reword this a little.  Say something like:

~~~
For files objects, specifying `newline` is not sufficient to set the line terminator.  You must pass in the line terminator explicitly, even in this case.
~~~

jreback commented:
@gfyoung you ok with this, can you rebase this? or @deflatSOCO

gfyoung commented:
@jreback : Can do, given that no reply how has come since the latest comment in > 1 week.

gfyoung committed:
BUG: Use correct line terminator on Windows

* Use OS line terminator if none is provided
* Enforce line terminator selection if one is

Originally authored by @deflatSOCO, but reapplied
by @gfyoung due to enormous merge conflicts.

Closes gh-20353.

gfyoung commented:
Alas, there were too many merge conflicts with `master` that I had to nuke the old commits and replace with it a squashed version.

gfyoung commented:
@jreback : Addressed all comments, rebased, and all is green.  PTAL.

jreback commented:
thanks @deflatSOCO and @gfyoung
